### PR TITLE
Nth fibonacci number using matrix exponentiation

### PR DESCRIPTION
**Pull request by villayatali123**
#### Checklist

- [ ] Added description of change
- [ ] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [ ] Added tests and example, test must pass
- [ ] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [ ] Relevant documentation/comments is changed or added
- [ ] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [ ] Search previous suggestions before making a new one, as yours may be a duplicate.
- [ ] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1215"></a>

### CODE REVIEW

**Review by nishant230701 — CHANGES_REQUESTED** [review]
why even use 
#include <bits/stdc++.h?>
Just use 
#include <iostream>
#include <vector>

File: math/fibonacci_matrix_exponentiation.cpp [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+ * @brief Computes N^th Fibonacci number given as
+ * input argument
+ *
+ * Takes O(logn) time to compute nth Fibonacci number
+ * 
+ *
+ * \author [villayatali123](https://github.com/villayatali123)
+ * \author [unknown author]()
+ * @see fibonacci.cpp, fibonacci_fast.cpp, string_fibonacci.cpp, fibonacci_large.cpp
+ */
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:12]:
```suggestion
```

File: math/fibonacci_matrix_exponentiation.cpp [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+/* @details
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:13]:
```suggestion
 * @details
```

File: math/fibonacci_matrix_exponentiation.cpp [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+ *Begin 
+ * Take two 2 dimensional vector
+ * Create a function and Perform matrix multiplication
+ * Ceate another function to find out power of matrix
+ * Create a function then to find out the Fibonacci number
+ * Multiply(arr1[2][2], arr2[2][2], int size)
+ * Take 4 variables a, b, c, d
+ * a = arr1[0][0] * arr2[0][0] + arr1[0][1] * arr2[1][0]
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:21]:
```suggestion
 * ```
 * a = arr1[0][0] * arr2[0][0] + arr1[0][1] * arr2[1][0]
```

**Review by kvedala — CHANGES_REQUESTED** [review]
1. Please review the contribution guidelines
2. no global variables
3. if need to use variables to track, you can do either of the following:
    1. use namespaces 
    2. use structs or classes
4. update documentation of the function similar to other files, [like this](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/math/fibonacci_large.cpp)

File: math/fibonacci_matrix_exponentiation.cpp [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+ * input argument
+ *
+ * Takes O(logn) time to compute nth Fibonacci number
+ * 
+ *
+ * \author [villayatali123](https://github.com/villayatali123)
+ * \author [unknown author]()
+ * @see fibonacci.cpp, fibonacci_fast.cpp, string_fibonacci.cpp, fibonacci_large.cpp
+ * @details
+ *Begin
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:13]: If you can review the checklist, one of the items is about the documentation. It explains what protocol we follow: doxygen. In this protocol, all documentations will be interpreted in Markdown syntax. These successive lines, though look good in the code, will not render properly i.e., they will not be rendered as WYSIWYG. To add breaks, add a new line "\n" at the end of each line. Please review the existing codes, references, checklist.

File: math/fibonacci_matrix_exponentiation.cpp [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+ * @see fibonacci.cpp, fibonacci_fast.cpp, string_fibonacci.cpp, fibonacci_large.cpp
+ * @details
+ *Begin 
+ * Take two 2 dimensional vector
+ * Create a function and Perform matrix multiplication
+ * Ceate another function to find out power of matrix
+ * Create a function then to find out the Fibonacci number
+ * Multiply(arr1[2][2], arr2[2][2], int size)
+ * Take 4 variables a, b, c, d
+ * '''
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:20]: as suggested before, the character to mark a verbatim section is: [`&grave;`](https://www.toptal.com/designers/htmlarrows/punctuation/grave-accent/) and not [`'`](https://www.toptal.com/designers/htmlarrows/punctuation/apostrophe/)

**Review by kvedala — CHANGES_REQUESTED** [review]
I suggest read the guidelines and the checklist at least once.

File: math/fibonacci_matrix_exponentiation.cpp [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+#include<iostream>
+#include<vector>
+#include <cassert>
+
+/**
+ * This function finds nth fibonacci number.
+ * @param n nth fibonacci number
+ * @param mod  modulo number to avoid overflow
+ */
+int fibo(int n , int mod )
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:23]:
```suggestion
uint64_t fibo(uint64_t n , uint64_t mod )
```
to handle large numbers

File: math/fibonacci_matrix_exponentiation.cpp [hunk 7]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @file fibonacci_matrix_exponentiation.cpp
+ * @brief Computes N^th Fibonacci number given as
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:3]: Not true. This program computes the "N^th Fibonacci number in modulo M" which is a bit different.

File: math/fibonacci_matrix_exponentiation.cpp [hunk 8]
```diff
@@ -0,0 +1,109 @@
+/**
+ * @file fibonacci_matrix_exponentiation.cpp
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:2]:
```suggestion
 * @file 
```

File: math/fibonacci_matrix_exponentiation.cpp [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+ * \author [unknown author]()
+ * @see fibonacci.cpp, fibonacci_fast.cpp, string_fibonacci.cpp, fibonacci_large.cpp
+ */
+
+#include<iostream>
+#include<vector>
+#include <cassert>
+
+/**
+ * This function finds nth fibonacci number.
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:19]:
```suggestion
 * This function finds nth fibonacci number in a given modulus.
```

File: math/fibonacci_matrix_exponentiation.cpp [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+ * @param n nth fibonacci number
+ * @param mod  modulo number to avoid overflow
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:21]:
```suggestion
 * @param mod  modulo number
```
the Fibonacci series in modulus has its own applications.

File: math/fibonacci_matrix_exponentiation.cpp [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+ */
+int fibo(int n , int mod )
+{
+	std::vector<int> result(2,0);
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:25]: wherever there are large unsigned integers, use `uint64_t`.

File: math/fibonacci_matrix_exponentiation.cpp [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+ * Function to test above algorithm
+ */
+void test()
+{
+    assert(fibo(6, 1000000007 ) == 8);
+    std::cout << "test case:1 passed\n";
+    assert(fibo(5, 1000000007  ) == 5);
+    std::cout << "test case:2 passed\n";
+    assert(fibo(10 , 1000000007) == 55);
+    std::cout << "test case:3 passed\n";
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:95]: can also add some other cases for different modulus:
Fibonacci(500, 100)
Fibonacci(500, 1000)
Fibonacci(50, 100)
Fibonacci(50, 200)

File: math/fibonacci_matrix_exponentiation.cpp [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+/**
+ * Main function
+ */
+int main()
+{
+	test();
+	int mod=1000000007;
+	std::cout<<"Enter the value of N: ";
+	int n=0; std::cin>>n; 
+	std::cout<<n<<"th Fibonacci number :"<<fibo( n , mod);
```
Comment (kvedala) [inline math/fibonacci_matrix_exponentiation.cpp:108]:
```suggestion
	std::cout<<n<<"th Fibonacci number in modulo " << mod << ": "<< fibo( n , mod) << std::endl; 
```

**Review by kvedala — CHANGES_REQUESTED** [review]
👍

**Comment by villayatali123** [comment]
i have made the changes please review it .

**Comment by villayatali123** [comment]
Approve it ASAP !

**Comment by villayatali123** [comment]
Approve it please ASAP !

**Review by kvedala — APPROVED** [review]

**Pull request merged**